File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n=s.size();
        bool isEven=(n%2==0);
        int maxlen=0;
        string res="";
        int l=0;
        int r=0;
        for(int i=0;i<n;i++){
            //ODD
            l=i;
            r=i;
            while((l>=0 && r<=n-1) && (s[l]==s[r])){
                if((r-l+1)>=maxlen){
                    maxlen=r-l+1;
                    res=s.substr(l,r-l+1);
                    cout<<res<<endl;
                }
                
                l--;
                r++;       
            }

            l=i;
            r=i+1;
            while((l>=0 && r<=n-1) && (s[l]==s[r])){
                if((r-l+1)>=maxlen){
                    maxlen=r-l+1;
                    res=s.substr(l,r-l+1);   
                }
                
                l--;
                r++;       
            }

        }
        
        cout<<" l:"<<l<<" r:"<<r<<endl;
        if(res==""){
            res=s[0];
        }
        return res;
    }
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        
        maxIdx=0
        maxLen=1
        n=len(s)
                
            
        def expandSearch(i,j):
            while (i>=0 and j<n and s[i]==s[j]):
                i-=1
                j+=1
            return (j-i+1-2,i+1)
    
        for idx in range(n):
            l1,s1=expandSearch(idx,idx)
            l2,s2=expandSearch(idx,idx+1)
            maxLen=max(maxLen,l1,l2)
            if(maxLen==l1):
                maxIdx=s1
            elif (maxLen==l2):
                maxIdx=s2
                
        return s[maxIdx:maxIdx+maxLen]